File: api/go/transactions.go
package handler

import (
	"net/http"
	"strconv"
	
	"github.com/budget-buddy/api/lib"
)

// Handler handles transaction CRUD operations
func Handler(w http.ResponseWriter, r *http.Request) {
	config := lib.Config{
		RequireAuth:    true,
		AllowedMethods: []string{"GET", "POST", "PUT", "DELETE"},
		EnableCORS:     true,
	}

	handler := lib.CreateHandler(transactionHandler, config)
	handler(w, r)
}

func transactionHandler(w http.ResponseWriter, r *http.Request) {
	user, ok := lib.GetUserFromContext(r)
	if !ok {
		lib.ErrorResponse(w, "Unauthorized", http.StatusUnauthorized, nil)
		return
	}

	switch r.Method {
	case "GET":
		handleGetTransactions(w, r, user)
	case "POST":
		handleCreateTransaction(w, r, user)
	case "PUT":
		handleUpdateTransaction(w, r, user)
	case "DELETE":
		handleDeleteTransaction(w, r, user)
	default:
		lib.ErrorResponse(w, "Method not allowed", http.StatusMethodNotAllowed, nil)
	}
}

func handleGetTransactions(w http.ResponseWriter, r *http.Request, user *lib.User) {
	// Parse query parameters
	limitStr := lib.GetQueryParam(r, "limit", "50")
	offsetStr := lib.GetQueryParam(r, "offset", "0")
	transactionType := lib.GetQueryParam(r, "type", "")
	category := lib.GetQueryParam(r, "category", "")

	limit, _ := strconv.Atoi(limitStr)
	offset, _ := strconv.Atoi(offsetStr)

	// TODO: Query database
	// For now, return mock data
	transactions := []map[string]interface{}{
		{
			"id":          "trans-1",
			"user_id":     user.ID,
			"amount":      100.50,
			"category":    "Groceries",
			"type":        "expense",
			"description": "Weekly shopping",
			"date":        "2024-01-15T10:00:00Z",
		},
	}

	// Apply filters
	var filtered []map[string]interface{}
	for _, t := range transactions {
		if transactionType != "" && t["type"] != transactionType {
			continue
		}
		if category != "" && t["category"] != category {
			continue
		}
		filtered = append(filtered, t)
	}

	// Calculate summary
	summary := map[string]interface{}{
		"totalIncome":   0.0,
		"totalExpenses": 100.50,
		"count":         len(filtered),
	}

	response := map[string]interface{}{
		"transactions": filtered,
		"summary":      summary,
		"pagination": map[string]interface{}{
			"total":   len(filtered),
			"limit":   limit,
			"offset":  offset,
			"hasMore": false,
		},
	}

	lib.SuccessResponse(w, response, http.StatusOK)
}

func handleCreateTransaction(w http.ResponseWriter, r *http.Request, user *lib.User) {
	var input lib.CreateTransactionInput
	if err := lib.ParseJSONBody(r, &input); err != nil {
		lib.ErrorResponse(w, "Invalid JSON body", http.StatusBadRequest, map[string]string{
			"error": err.Error(),
		})
		return
	}

	// Validate input
	if input.Amount <= 0 {
		lib.ErrorResponse(w, "Amount must be positive", http.StatusBadRequest, nil)
		return
	}

	if input.Category == "" {
		lib.ErrorResponse(w, "Category is required", http.StatusBadRequest, nil)
		return
	}

	if input.Type != "income" && input.Type != "expense" {
		lib.ErrorResponse(w, "Type must be 'income' or 'expense'", http.StatusBadRequest, nil)
		return
	}

	// TODO: Insert into database
	transaction := map[string]interface{}{
		"id":             "trans-new",
		"user_id":        user.ID,
		"amount":         input.Amount,
		"category":       input.Category,
		"type":           input.Type,
		"description":    input.Description,
		"date":           input.Date,
		"merchant":       input.Merchant,
		"payment_method": input.PaymentMethod,
		"created_at":     "2024-01-15T10:00:00Z",
	}

	lib.SuccessResponse(w, map[string]interface{}{
		"transaction": transaction,
	}, http.StatusCreated)
}

func handleUpdateTransaction(w http.ResponseWriter, r *http.Request, user *lib.User) {
	id := lib.GetQueryParam(r, "id", "")
	if id == "" {
		lib.ErrorResponse(w, "Transaction ID required", http.StatusBadRequest, nil)
		return
	}

	var input map[string]interface{}
	if err := lib.ParseJSONBody(r, &input); err != nil {
		lib.ErrorResponse(w, "Invalid JSON body", http.StatusBadRequest, nil)
		return
	}

	// TODO: Update in database
	transaction := map[string]interface{}{
		"id":      id,
		"user_id": user.ID,
		"updated": true,
	}

	lib.SuccessResponse(w, map[string]interface{}{
		"transaction": transaction,
	}, http.StatusOK)
}

func handleDeleteTransaction(w http.ResponseWriter, r *http.Request, user *lib.User) {
	id := lib.GetQueryParam(r, "id", "")
	if id == "" {
		lib.ErrorResponse(w, "Transaction ID required", http.StatusBadRequest, nil)
		return
	}

	// TODO: Delete from database

	lib.SuccessResponse(w, map[string]interface{}{
		"message": "Transaction deleted successfully",
		"id":      id,
	}, http.StatusOK)
}

File: api/transactions.ts
/**
 * Serverless Function: Transactions API
 * Handles CRUD operations for user transactions
 */

import { VercelRequest, VercelResponse } from '@vercel/node';
import {
  createServerlessHandler,
  AuthenticatedRequest,
  validateMethod,
  parseBody,
  successResponse,
  errorResponse,
  getSupabaseClient,
} from './_lib/serverless-helpers';

async function handler(req: AuthenticatedRequest, res: VercelResponse): Promise<void> {
  const supabase = getSupabaseClient(req);

  // GET: Fetch transactions
  if (req.method === 'GET') {
    const { searchParams } = new URL(req.url!, `http://${req.headers.host}`);
    const limit = parseInt(searchParams.get('limit') || '50');
    const offset = parseInt(searchParams.get('offset') || '0');
    const type = searchParams.get('type'); // 'income' or 'expense'

    let query = supabase
      .from('transactions')
      .select('*', { count: 'exact' })
      .eq('user_id', req.user!.id)
      .order('date', { ascending: false })
      .range(offset, offset + limit - 1);

    if (type) {
      query = query.eq('type', type);
    }

    const { data, error, count } = await query;

    if (error) {
      errorResponse(res, 'Failed to fetch transactions', 500, error);
    }

    successResponse(res, {
      transactions: data,
      total: count,
      limit,
      offset,
    });
  }

  // POST: Create transaction
  if (req.method === 'POST') {
    const body = await parseBody(req);

    if (!body || !body.amount || !body.category || !body.type) {
      errorResponse(res, 'Missing required fields', 400);
    }

    const { data, error } = await supabase
      .from('transactions')
      .insert({
        user_id: req.user!.id,
        amount: body.amount,
        category: body.category,
        type: body.type,
        description: body.description || '',
        date: body.date || new Date().toISOString(),
        merchant: body.merchant || '',
        payment_method: body.payment_method || '',
      })
      .select()
      .single();

    if (error) {
      errorResponse(res, 'Failed to create transaction', 500, error);
    }

    successResponse(res, { transaction: data }, 201);
  }

  // PUT: Update transaction
  if (req.method === 'PUT') {
    const body = await parseBody(req);
    const { searchParams } = new URL(req.url!, `http://${req.headers.host}`);
    const id = searchParams.get('id');

    if (!id) {
      errorResponse(res, 'Transaction ID required', 400);
    }

    const { data, error } = await supabase
      .from('transactions')
      .update(body)
      .eq('id', id)
      .eq('user_id', req.user!.id)
      .select()
      .single();

    if (error) {
      errorResponse(res, 'Failed to update transaction', 500, error);
    }

    successResponse(res, { transaction: data });
  }

  // DELETE: Delete transaction
  if (req.method === 'DELETE') {
    const { searchParams } = new URL(req.url!, `http://${req.headers.host}`);
    const id = searchParams.get('id');

    if (!id) {
      errorResponse(res, 'Transaction ID required', 400);
    }

    const { error } = await supabase
      .from('transactions')
      .delete()
      .eq('id', id)
      .eq('user_id', req.user!.id);

    if (error) {
      errorResponse(res, 'Failed to delete transaction', 500, error);
    }

    successResponse(res, { message: 'Transaction deleted' });
  }

  errorResponse(res, 'Method not allowed', 405);
}

export default createServerlessHandler(handler, {
  auth: true,
  cors: true,
});
